Add URL query string tools

roman-mazur opened the pull request:
AnalyticsUrlParam class encapsulates the logic of handling
passed URL parameters with anonymous device identifier.

Change-type: minor
Signed-off-by: Roman Mazur <roman@balena.io>

roman-mazur committed:
Add URL query string tools

AnalyticsUrlParam class encapsulates the logic of handling
passed URL parameters with anonymous device identifier.

Change-type: minor
Signed-off-by: Roman Mazur <roman@balena.io>

nazrhom commented:
@balena-ci retest

roman-mazur commented:
@balena-ci retest please

roman-mazur commented:
@balena-ci retest please

roman-mazur commented:
@balena-ci retest please

webpack output is re-configured

roman-mazur commented:
Example of the compiled bundle served via unpkg:
https://unpkg.com/analytics-client@0.1.0-roman-url-params-431f6998b0134ff6209577529d1d3737fc8b1bbe/dist/bundle.js

sradevski commented on webpack.config.js:
```diff
@@ -0,0 +1,24 @@
+const path = require('path');
+
+module.exports = {
```
Can we achieve the same only using the typescript compiler? Why do we need to use webpack?

roman-mazur replied:
We are using webpack to create a `bundle.js` (that contains all necessary pieces of this library) loadable by our blog and docs (via unpkg).

sradevski replied:
There is also `--outFile` option for the typescript compiler, but I just read there are some gotchas with it, so webpack might be the way to go.

sradevski commented on package.json:
```diff
@@ -0,0 +1,36 @@
+    "lodash": "^4.17.15",
+    "typescript": "^3.6.3"
+  },
+  "devDependencies": {
```
Can you add `lint-staged` and `husky` so we run linting and prettifying on commit?

roman-mazur replied:
On it...

roman-mazur replied:
Added.

sradevski commented on src/url-params.ts:
```diff
@@ -0,0 +1,111 @@
+
+const deviceIdSeparator = /\s*,\s*/;
+
+const cookiesSupported = hasCookieSupport();
```
Maybe it's better to run this inside the constructor. I guess the environment won't change after loading the script, but it's better if the check is delayed only when the consumer decides they want to initialize the script.

roman-mazur replied:
Moved to the constructor.

sradevski commented on src/url-params.ts:
```diff
@@ -0,0 +1,111 @@
+
+		const passedDeviceId = params.get(URL_PARAM_DEVICE_ID);
+		if (passedDeviceId) {
+			let originalMpId: string | null = null;
```
I would prefer if this is a full name `originalMixpanelId` as `Mp` can be confusing without thinking about it.

roman-mazur replied:
Renamed.

sradevski commented on src/url-params.ts:
```diff
@@ -0,0 +1,111 @@
+	consumeUrlParameters(queryString: string): string | null {
+		const params = new URLSearchParams(queryString);
+
+		const passedDeviceId = params.get(URL_PARAM_DEVICE_ID);
```
I would suggest parsing the parameter here from `string` to `string[]`, remove `takeFirstId` as it is a bit pointless at that point, push the original Mixpanel ID here, and pass only the parsed array to `storeInboundDeviceIds`. I feel like this separates the concerns of each function more clearly.

roman-mazur replied:
PTAL at the current implementation.

sradevski commented on src/url-params.ts:
```diff
@@ -0,0 +1,111 @@
+
+	constructor(private mixpanel?: Mixpanel) {}
+
+	private storeInboundDeviceIds(ids: string, currentDeviceId: string | null) {
```
```suggestion
	private setDeviceIds(ids: string, currentDeviceId: string | null) {
```

roman-mazur replied:
Renamed.

sradevski commented on src/url-params.ts:
```diff
@@ -0,0 +1,111 @@
+		return res;
+	}
+
+	private retrieveAndMerge(input: string[]) {
```
If you do what I suggested in the constructor, this won't be necessary (you can just use `this.deviceIds` directly). Then, the caller will turn it into a string if they wish to.

roman-mazur replied:
This function is removed now.

sradevski commented on src/url-params.ts:
```diff
@@ -0,0 +1,111 @@
+export class AnalyticsUrlParams {
+	private deviceIds: string | null;
+
+	constructor(private mixpanel?: Mixpanel) {}
```
How about you check here if there are deviceIds in cookies, and set the class `deviceIds` from there? Like that you can only reference the class property, and only the cookies when you are setting.

roman-mazur replied:
Check the current impl.

sradevski commented on src/url-params.ts:
```diff
@@ -0,0 +1,111 @@
+	/**
+	 * @return all anonymous device IDs that can be passed to other sites, separated with comma
+	 */
+	allDeviceIds(): string {
```
I think it is better to return a `string[]` instead of the already concatenated string. You can have a separate method that will do the concatenation, otherwise the name is confusing (judging by the name, I would expect an array).

roman-mazur replied:
Yep, I agree.
this method now returns an array

sradevski commented on src/url-params.ts:
```diff
@@ -0,0 +1,111 @@
+	 */
+	allDeviceIds(): string {
+		if (this.deviceIds == null) {
+			const knownIds = this.mixpanel == null ? [] : [ this.mixpanel.get_distinct_id() ];
```
Can we also do this in the constructor? If so, we can have `getDeviceIds` function that will simply be a getter for `this.deviceids`.

roman-mazur replied:
It now returns the deviceId property value.

sradevski commented on src/url-params.ts:
```diff
@@ -0,0 +1,111 @@
+	/**
+	 * @return part of a query parameter string that can be appended to URLs
+	 */
+	deviceIdQuery(): string {
```
```suggestion
	getDeviceIdsQueryString(): string {
```

roman-mazur replied:
Renamed.

sradevski reviewed (commented):
I think the url params class can be simplified, as currently there are a couple sources of truth (cookies, the local `deviceIds` and some functions append the mixpanel id to the result. If something is confusing, let me know and we can go through it together.

roman-mazur commented:
@sradevski your comments should be addressed now.

roman-mazur commented:
@balena-ci retest please

roman-mazur commented:
@balena-ci retest please

roman-mazur commented:
@balena-ci retest again, please
both TS and webpack outputs are now in the dist folder

roman-mazur committed:
Avoid calling mixpanel in constructor

It's possible clients will construct AnalyticsUrlParams before
mixpanel is initialized.

Change-type: patch
Signed-off-by: Roman Mazur <roman@balena.io>

roman-mazur commented:
@balena-ci retest please, again

sradevski commented on README.md:
```diff
@@ -1,2 +1,24 @@
+
+analytics.consumeUrlParameters(window.location.search);
+
+const signupUrl = '/signup?' + analytics.deviceIdQuery();
```
Update this with the updated method name.

roman-mazur replied:
updated

sradevski commented on src/url-params.ts:
```diff
@@ -0,0 +1,105 @@
+	 * @return all anonymous device IDs that can be passed to other sites
+	 */
+	allDeviceIds() {
+		const mixpanelId = this.mixpanel ? [this.mixpanel.get_distinct_id()] : [];
```
I think we can also do this check and assignment in the constructor, since that's where we get mixpanel passed. In that case we can have this method to be `getAllDeviceIds` and it will simply return `this.deviceIds`.

roman-mazur replied:
See https://github.com/balena-io/analytics-client/pull/1/commits/bc5b491cad22dd7a8de7498a68b5772046bf4285

What you describe was my initial implementation, but then I changed it in the commit above making sure we don't expect mixpanel library to be initialized in the constructor.

sradevski replied:
If mixpanel is not passed in the constructor, there is no way to update it afterwards, as it is private. If a mixpanel client is indeed passed to the constructor, but it is not ready yet, then `this.mixpanel` will always return true, so this check will be wrong. Maybe I am missing something?

roman-mazur replied:
A typical use of mixpanel:
```
import mixpanel from 'mixpanel-browser';

// AnalyticsUrlParams can be initialized at this moment of time.

mixpanel.init(token, options);

// Only now you can
mixpanel.get_distinct_id();
```

So we can do `const analyticsTool = new AnalyticsUrlParams(mixpanel)`, but `init ` is not yet called on the mixpanel object.

sradevski replied:
But `mixpanel` is never falsy, so `this.mixpanel ? ... : ...` will always be true. I think we need to do something like `this.mixpanel.isInitialized` if there is such thing.

roman-mazur replied:
The constructor is declared in a way it allows the mixpanel object to be skipped.
The logic handles this particular case (so it works when we do `new AnalyticsUrlParams()`).
The main point is that `get_distinct_id` call should be delayed/lazy, which is why `allDeviceIds` method is a bit more complicated.

Speaking about `isInitialized`, there is [no such method](https://github.com/mixpanel/mixpanel-js/issues/133) in mixpanel.

sradevski commented on src/url-params.ts:
```diff
@@ -0,0 +1,105 @@
+			const originalMixpanelId =
+				this.mixpanel != null ? this.mixpanel.get_distinct_id() : null;
+
+			const newCurrentDeviceId = this.setDeviceIds(
```
Normally you wouldn't expect a setter to return anything, and in this case returning the first element is a bit confusing. How about 
```
this.setDeviceIds(passedDeviceIds);
const newCurrentDeviceId = this.deviceIds ? this.deviceIds[0] : null;
```

If you set `this.deviceIds` to the mixpanel distinct ID in the constructor, `setDeviceIds` will only require a single parameter, as `mixpanelId` will already by in `deviceIds`.

roman-mazur replied:
Please note that this is not a public setter, but a private method and an implementation detail.

If we remove parsing of the input string, we'll need to parse it separately in the constructor (because cookie value is a string) and in `consumeUrlParameters`.
Since it does the parsing, it also returns the first device ID that can be used to configure mixpanel.

That said, we can rename it so it's not confused with a setter semantics. To something like `updateDeviceIds`.

sradevski replied:
I think whether a function is public or private doesn't matter much in terms of functions being intuitive in their behavior. In this case, you will just read from `this.deviceIds` after calling `setDeviceIds` instead of expecting a return parameter, that's all.

roman-mazur replied:
>I think whether a function is public or private doesn't matter much in terms of functions being intuitive in their behavior

My main point is not about public/private, but about the fact that it's not necessarily a classic setter. 

Also, although `this.deviceId` is an array, it's really a set of device IDs: when setDeviceIds is called, we merge the existing list with a new one, removing duplicates. The order of elements is not really guaranteed, hence, we should not assume `this.deviceIds[0]` is the first ID from the URL query.

sradevski reviewed (commented):
A couple more comments to further simplify the code, the rest is 👍

roman-mazur commented:
@balena-ci retest please

roman-mazur committed:
Remove lodash dependency

roman-mazur commented:
lodash dependency is removed now, `Set` is used to maintain device IDs.

@balena-ci retest, please
